{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deconvolution with a panel of methods\n",
    "\n",
    "We compare the predictive value of CLIMB and a panel of published bulk deconvolution methods: BayesPrism, CiberSortX, MUSIC, BisqueRNA, and NNLS, TAPE, Scaden, and using random proportions as a negative control.\n",
    "\n",
    "*Note: As CiberSortX deconvolution was obtained via their website interface, we provide the code to produce the tables given as input for CiberSortX, and use here the table containing the processed results of cell type proportions.*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(glmnet) ; library(Biobase) ; library(reshape2) ; library(BayesPrism)\n",
    "library(MuSiC) ; library(SingleCellExperiment) ; library(BisqueRNA)\n",
    "library(climb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo alexdray86/climb@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking for file ‘/private/var/folders/z6/0fzn0d4s1v74_s44tjrc9mg00000gp/T/Rtmp5MYwVx/remotes12f2a473e8afa/alexdray86-CLIMB-73ef455/DESCRIPTION’\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mpreparing ‘climb’:\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[32m✔\u001b[39m  \u001b[90mchecking DESCRIPTION meta-information\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for LF line-endings in source and make files and shell scripts\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mchecking for empty or unneeded directories\u001b[39m\u001b[36m\u001b[39m\n",
      "\u001b[90m─\u001b[39m\u001b[90m  \u001b[39m\u001b[90mbuilding ‘climb_0.0.5.tar.gz’\u001b[39m\u001b[36m\u001b[39m\n",
      "   \n",
      "\r"
     ]
    }
   ],
   "source": [
    "### UPDATE CLIMB ###\n",
    "detach(\"package:climb\", unload=TRUE)\n",
    "devtools::install_github('alexdray86/climb', force=T) # Need force=T if already installed ! \n",
    "library(climb)\n",
    "# You can check with sessionInfo() that climb is version 0.0.5 (should be climb_0.0.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# METRICS FUNCTIONS\n",
    "num <- function(x){ return(as.numeric(as.character(x)))}\n",
    "flatten <- function(x){return(as.vector(as.matrix(x)))}\n",
    "mse <- function(tr, pr){ return( sum((tr - pr)^2) / length(tr) ) }\n",
    "rmse <- function(tr, pr){ return( sqrt( sum((tr - pr)^2) / length(tr) ) ) }\n",
    "scor <- function(tr, pr){ return( cor(tr, pr, method = 'spearman') ) }\n",
    "pcor <- function(tr, pr){ return( cor(tr, pr, method = 'pearson') ) }\n",
    "r_square <- function(pr, tr){\n",
    "    SStot = sum((tr - mean(tr))^2)\n",
    "    SSres = sum((pr - tr)^2)\n",
    "    return(1 - SSres / SStot)\n",
    "}\n",
    "r_square <- function(pr, tr){\n",
    "    return( sign(cor(pr,tr))*cor(pr,tr)^2 )\n",
    "}\n",
    "### Function to retrieve all metrics from Deconvolution vs True cell-type proportion comparison\n",
    "comp_rows_and_cols <- function(pred_prop, true_prop){\n",
    "    pc_l = list() ; sc_l = list()\n",
    "    rsq_l = list(); rmse_l = list()\n",
    "    for(i in 1:dim(pred_prop)[1]){\n",
    "        pc_l[[i]] = cor(pred_prop[i,], true_prop[i,], method = 'pearson')\n",
    "        sc_l[[i]] = cor(pred_prop[i,], true_prop[i,], method = 'spearman')\n",
    "        rsq_l[[i]] = max(r_square(pred_prop[i,], true_prop[i,]), -1.0)\n",
    "        rmse_l[[i]] = rmse(pred_prop[i,], true_prop[i,])\n",
    "    }\n",
    "    pc_c = list() ; sc_c = list()\n",
    "    rsq_c = list(); rmse_c = list()\n",
    "    for(i in 1:dim(pred_prop)[2]){\n",
    "        pc_c[[i]] = cor(pred_prop[,i], true_prop[,i], method = 'pearson')\n",
    "        sc_c[[i]] = cor(pred_prop[,i], true_prop[,i], method = 'spearman')\n",
    "        rsq_c[[i]] = max(r_square(pred_prop[,i], true_prop[,i]), -1.0)\n",
    "        rmse_c[[i]] = rmse(pred_prop[,i], true_prop[,i])\n",
    "    }\n",
    "    res = list() ; \n",
    "    res[[1]] = num(pc_l) ; res[[2]] = num(sc_l)\n",
    "    res[[3]] = num(rsq_l); res[[4]] = num(rmse_l)\n",
    "    res[[5]] = num(pc_c) ; res[[6]] = num(sc_c)\n",
    "    res[[7]] = num(rsq_c); res[[8]] = num(rmse_c)\n",
    "    res[[9]] = pcor(as.vector(num(pred_prop)), as.vector(num(true_prop)))\n",
    "    res[[10]] = scor(as.vector(num(pred_prop)), as.vector(num(true_prop)))\n",
    "    res[[11]] = max(r_square(as.vector(num(pred_prop)), as.vector(num(true_prop))), -1.0)\n",
    "    res[[12]] = rmse(as.vector(num(pred_prop)), as.vector(num(true_prop)))\n",
    "    return(res)\n",
    "}\n",
    "reformat_strings <- function(vector_string){\n",
    "    # replace plus and minus (e.g. useful for CD34+, CD34- populations)\n",
    "    vector_string <- gsub('\\\\-$', 'minus', vector_string) \n",
    "    vector_string <- gsub('\\\\+', 'plus', vector_string) \n",
    "    vector_string <- gsub('\\\\ ', '\\\\.', vector_string) \n",
    "    # remove all special characters\n",
    "    vector_string <- gsub('[^[:alnum:] ]','',vector_string)\n",
    "    return(vector_string)\n",
    "}\n",
    "reformat_celltypes <- function(celltype_labels){\n",
    "    celltype_labels <- reformat_strings(as.vector(celltype_labels))\n",
    "    celltype_labels <- factor(celltype_labels)\n",
    "    return(celltype_labels)\n",
    "}\n",
    "add_missing_celltypes <- function(tab, celltype_names){\n",
    "    missing_ct = celltype_names[!celltype_names %in% colnames(tab)]\n",
    "    if(length(missing_ct)>0){\n",
    "        for(ct in missing_ct){\n",
    "            tab_ = cbind(tab, rep(0, dim(tab)[1]))\n",
    "            colnames(tab_) = c(colnames(tab), ct)\n",
    "            tab = tab_\n",
    "        }\n",
    "    }\n",
    "    return(tab)\n",
    "} \n",
    "\n",
    "### Function for each deconvolution methods ### \n",
    "deconvolution_climb <- function(sc, bulks, true_prop, celltype_names){\n",
    "    climb_res = climb(sc, bulks)\n",
    "    climb_prop = climb_res$props\n",
    "    rownames(climb_prop) = colnames(bulks)\n",
    "    climb_prop = add_missing_celltypes(climb_prop, celltype_names)\n",
    "    return(as.matrix(climb_prop[,celltype_names]))\n",
    "}\n",
    "deconvolution_bayesprism <- function(sc, bulks, true_prop, celltype_names){\n",
    "    myPrism <- new.prism(reference= t(exprs(sc)), \n",
    "                         mixture=t(exprs(bulks)), input.type=\"count.matrix\", \n",
    "                         cell.type.labels = as.character(sc$cellType), \n",
    "                         cell.state.labels = as.character(sc$cellType), \n",
    "                         key = NULL,outlier.cut=0.01,outlier.fraction=0.1)\n",
    "    bp.res <- run.prism(prism = myPrism, n.cores=6)\n",
    "    bayesprism_prop <- get.fraction (bp=bp.res, which.theta=\"final\",\n",
    "             state.or.type=\"type\")\n",
    "    bayesprism_prop = add_missing_celltypes(bayesprism_prop, celltype_names)\n",
    "    return(as.matrix(bayesprism_prop[,celltype_names]))\n",
    "}\n",
    "deconvolution_music <- function(sc, bulks, true_prop, celltype_names){\n",
    "    sce <- SingleCellExperiment(list(counts=exprs(sc)),\n",
    "                                colData=DataFrame(label=colnames(sc),\n",
    "                                                  cellType=sc$cellType,\n",
    "                                                  SubjectName=sc$SubjectName),\n",
    "                                rowData=DataFrame(gene_name=rownames(sc)))\n",
    "    music.output = music_prop(bulk.mtx = exprs(bulks), sc.sce = sce, \n",
    "                                 clusters = 'cellType',samples = 'SubjectName',verbose = F)\n",
    "    music_prop = music.output$Est.prop.weighted\n",
    "    music_prop = add_missing_celltypes(music_prop, celltype_names)\n",
    "    return(music_prop[,celltype_names])\n",
    "}\n",
    "deconvolution_nnls <- function(sc, bulks, true_prop, celltype_names){\n",
    "    sce <- SingleCellExperiment(list(counts=exprs(sc)),\n",
    "                                colData=DataFrame(label=colnames(sc),\n",
    "                                                  cellType=sc$cellType,\n",
    "                                                  SubjectName=sc$SubjectName),\n",
    "                                rowData=DataFrame(gene_name=rownames(sc)))\n",
    "    music.output = music_prop(bulk.mtx = exprs(bulks), sc.sce = sce, \n",
    "                                 clusters = 'cellType',samples = 'SubjectName',verbose = F)\n",
    "    nnls_prop = music.output$Est.prop.allgene\n",
    "    nnls_prop = add_missing_celltypes(nnls_prop, celltype_names)\n",
    "    return(nnls_prop[,celltype_names])\n",
    "}\n",
    "deconvolution_bisque <- function(sc, bulks, true_prop, celltype_names){\n",
    "    bisque.out = BisqueRNA::ReferenceBasedDecomposition(bulks, sc, use.overlap=F)\n",
    "    bisque_prop = t(bisque.out$bulk.props)\n",
    "    bisque_prop = add_missing_celltypes(bisque_prop, celltype_names)\n",
    "    return(bisque_prop[,celltype_names])\n",
    "}\n",
    "deconvolution_random <- function(nnls_prop){\n",
    "    nnls_prop = read.csv(paste0(out_dir, '/nnls_prop.csv'), row.names='X')\n",
    "    set.seed(1)\n",
    "    rand_prop = nnls_prop[sample(rownames(nnls_prop)), sample(colnames(nnls_prop))]\n",
    "    rownames(rand_prop) = rownames(nnls_prop) ; colnames(rand_prop) = colnames(nnls_prop)\n",
    "    return(rand_prop)\n",
    "}\n",
    "write_tables_cibersortx <- function(sc, bulks, true_prop, celltype_names, tab_dir){\n",
    "    # WARNING : the following script only writes tables to be used on CiberSortX website\n",
    "    sc_mat = exprs(sc)\n",
    "    colnames(sc_mat) = as.character(sc$cellType) # cell-type labels provided as column names\n",
    "    write.table(sc_mat , paste0(tab_dir,\"/TMP_cibersortx_sc_mat.txt\"), sep = '\\t', quote=F)\n",
    "    write.table(exprs(bulks), paste0(tab_dir,\"/TMP_cibersortx_pbulks.txt\"), sep = '\\t', quote=F)\n",
    "    # tables needs to be further processed in bash to add 'Gene' at the start of the file\n",
    "    # cat <(echo -ne \"Gene\\t\") TMP_cibersortx_pbulks.txt > cibersortx_pbulks.txt\n",
    "    # cat <(echo -ne \"Gene\\t\") TMP_cibersortx_sc_mat.txt > cibersortx_sc_mat.txt\n",
    "}\n",
    "write_tables_tape_scaden <- function(sc, bulks, true_prop, celltype_names, tab_dir){\n",
    "    # Write all input tables formatted for TAPE / Scaden\n",
    "    write.csv(t(exprs(bulks)), paste0(tab_dir, \"/tape_scaden_bulks_expr.csv\"), quote=F)\n",
    "    sc_mat_TAPE = exprs(sc)\n",
    "    sc_mat_TAPE = t(sc_mat_TAPE)\n",
    "    rownames(sc_mat_TAPE) = sc$cellType\n",
    "    write.csv(sc_mat_TAPE, paste0(tab_dir, \"/tape_scaden_scmat_expr.csv\"), quote=F)\n",
    "}\n",
    "deconvolution_scaden <- function(sc, bulks, true_prop, celltype_names, tab_dir){\n",
    "    # WARNING: launching TAPE / Scaden from the command line induces figures popping up\n",
    "    # which blocks the process (figures have to be closed manually). \n",
    "    # To prevent these figures to be made, one can comment lines 103-109,115-119,128-131, \n",
    "    # and 172-176 in the file 'tape_scaden_env/lib/python3.7/site-packages/TAPE/utils.py'\n",
    "    message('Launch Scaden / TAPE deconvolution')\n",
    "    # Launch bash script that will launch environment and python script\n",
    "    system(paste0('bash launch_TAPE_and_Scaden.sh ', tab_dir, '/tape_scaden_scmat_expr.csv ',\n",
    "                  tab_dir, '/tape_scaden_bulks_expr.csv ',\n",
    "                  '../REVISION_1/data/celltype_expression/TAPE/input/GeneLength.txt ',\n",
    "                  tab_dir, '/'), wait=TRUE)\n",
    "}\n",
    "load_tape_prop <- function(tmp_dir, out_dir, celltype_names){\n",
    "    celltype_names = reformat_celltypes(celltype_names)\n",
    "    tape_prop = read.csv(paste0(tmp_dir,'/tape_a_prop.csv'), row.names = 'X')\n",
    "    colnames(tape_prop) = reformat_celltypes(colnames(tape_prop))\n",
    "    tape_prop = add_missing_celltypes(tape_prop, celltype_names)\n",
    "    tape_prop = tape_prop[,as.character(celltype_names)]\n",
    "    write.csv(tape_prop, paste0(out_dir,'/tape_prop.csv'), quote=F)\n",
    "    return(tape_prop)\n",
    "}\n",
    "load_scaden_prop <- function(tmp_dir, out_dir, celltype_names){\n",
    "    celltype_names = reformat_celltypes(celltype_names)\n",
    "    scaden_prop = read.csv(paste0(tmp_dir,'/scaden_prop.csv'), row.names = 'X')\n",
    "    colnames(scaden_prop) = reformat_celltypes(colnames(scaden_prop))\n",
    "    scaden_prop = add_missing_celltypes(scaden_prop, celltype_names)\n",
    "    scaden_prop = scaden_prop[,as.character(celltype_names)]\n",
    "    write.csv(scaden_prop, paste0(out_dir,'/scaden_prop.csv'), quote=F)\n",
    "    return(scaden_prop)\n",
    "}\n",
    "load_all_props <- function(odir, true_p, celltype_names, include_cibersortx=FALSE){\n",
    "    tdir = paste0(odir,'/tmp_tabs')\n",
    "    climb_p = read.csv(paste0(odir, '/climb_prop.csv'), row.names='X')\n",
    "    bayesprism_p = read.csv(paste0(odir, '/bayesprism_prop.csv'), row.names='X')\n",
    "    music_p = read.csv(paste0(odir, '/music_prop.csv'), row.names='X')\n",
    "    nnls_p = read.csv(paste0(odir, '/nnls_prop.csv'), row.names='X')\n",
    "    rand_p = read.csv(paste0(odir, '/rand_prop.csv'), row.names='X')\n",
    "    bisque_p = read.csv(paste0(odir, '/bisque_prop.csv'), row.names='X')\n",
    "    tape_p = load_tape_prop(tdir, odir, celltype_names)\n",
    "    scaden_p = load_scaden_prop(tdir, odir, celltype_names)\n",
    "    if(include_cibersortx){\n",
    "        csx_p = read.csv(paste0(odir, '/cibersortx_prop.csv'), row.names='X')\n",
    "        all_p = list(climb_dw_p, climb_p, scaden_p, tape_p, csx_p, bayesprism_p, music_p, nnls_p, bisque_p, rand_p)\n",
    "        names(all_p) = c('CLIMB', 'Scaden', 'TAPE', 'CiberSortX', 'BayesPrism', 'MuSiC', 'NNLS', 'BisqueRNA', 'Random')\n",
    "    } else {\n",
    "        all_p = list(as.matrix(climb_dw_p), as.matrix(climb_p), as.matrix(scaden_p), \n",
    "                     as.matrix(tape_p), as.matrix(bayesprism_p), as.matrix(music_p), \n",
    "                     as.matrix(nnls_p), as.matrix(bisque_p), as.matrix(rand_p))\n",
    "        names(all_p) = c('CLIMB', 'Scaden', 'TAPE', 'BayesPrism', 'MuSiC', 'NNLS', 'BisqueRNA', 'Random')\n",
    "    }\n",
    "    true_prop_cols = colnames(true_prop) ; true_prop_rows = rownames(true_prop)\n",
    "    message(paste0('Checking that all tables have the same cell-type labels: ',suppressWarnings(all(lapply(all_p, function(x) all(colnames(x) == true_prop_cols))))))\n",
    "    message(paste0('Checking that all tables has the same sample names: ',suppressWarnings(all(lapply(all_p, function(x) all(rownames(x) == true_prop_rows))))))\n",
    "    return(all_p)\n",
    "}                                                                                           \n",
    "make_all_metrics <- function(all_ps, t_prop){\n",
    "    all_rs = list()\n",
    "    for(p in 1:length(all_ps)){\n",
    "        name_p = names(all_ps)[p]\n",
    "        all_rs[[name_p]] = suppressWarnings(comp_rows_and_cols(all_ps[[p]], t_prop))\n",
    "    }\n",
    "    names(all_rs) = names(all_ps)\n",
    "    return(all_rs)\n",
    "}\n",
    "                                                                                                  \n",
    "\n",
    "launch_all_deconvolution_methods <- function(sc, bulks, true_prop, celltype_names, out_dir){ \n",
    "    # Create folders\n",
    "    message('Creating folders')\n",
    "    tmp_dir = paste0(out_dir, '/tmp_tabs')\n",
    "    dir.create(tmp_dir, showWarnings = FALSE, recursive = TRUE)\n",
    "    \n",
    "    ### Write tables for CiberSortX deconvolution ### \n",
    "    message('Writing tables for CiberSortX')\n",
    "    write_tables_cibersortx(sc, bulks, true_prop, celltype_names, tmp_dir)\n",
    "    \n",
    "    ### CLIMB deconvolution ###\n",
    "    message('CLIMB deconvolution')\n",
    "    climb_prop = deconvolution_climb(sc, bulks, true_prop, celltype_names)\n",
    "    write.csv(climb_prop, paste0(out_dir, '/climb_prop.csv'), quote=F)\n",
    "    message(paste0('CLIMB Accruacy as PCC: ', cor(flatten(true_prop), flatten(climb_prop))))\n",
    "    \n",
    "    ### BayesPrism deconvolution ###\n",
    "    message('BayesPrism deconvolution')\n",
    "    bayesprism_prop = deconvolution_bayesprism(sc, bulks, true_prop, celltype_names)\n",
    "    write.csv(bayesprism_prop, paste0(out_dir, '/bayesprism_prop.csv'), quote=F)\n",
    "    message(paste0('BayesPrism Accruacy as PCC: ', cor(flatten(true_prop), flatten(bayesprism_prop))))\n",
    "    \n",
    "    ### MUSIC deconvolution ###\n",
    "    message('MUSIC deconvolution')\n",
    "    music_ps = deconvolution_music(sc, bulks, true_prop, celltype_names)\n",
    "    write.csv(music_ps, paste0(out_dir, '/music_prop.csv'), quote=F)\n",
    "    message(paste0('MuSiC Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(music_ps)))))\n",
    "    \n",
    "    ### NNLS deconvolution ### \n",
    "    message('NNLS deconvolution')\n",
    "    nnls_prop = deconvolution_nnls(sc, bulks, true_prop, celltype_names)\n",
    "    write.csv(nnls_prop, paste0(out_dir, '/nnls_prop.csv'), quote=F)\n",
    "    message(paste0('NNLS Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(nnls_prop)))))\n",
    "    \n",
    "    ### BisqueRNA deconvolution ### \n",
    "    message('BisqueRNA deconvolution')\n",
    "    bisque_prop = deconvolution_bisque(sc, bulks, true_prop, celltype_names)\n",
    "    write.csv(bisque_prop, paste0(out_dir, '/bisque_prop.csv'), quote=F)\n",
    "    message(paste0('BisqueRNA Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(bisque_prop)))))\n",
    "\n",
    "    ### Deconvolution Scaden / TAPE \n",
    "    message('TAPE / Scaden deconvolution')\n",
    "    write_tables_tape_scaden(sc, bulks, true_prop, celltype_names, tmp_dir)\n",
    "    deconvolution_scaden(sc, bulks, true_prop, celltype_names, tmp_dir)\n",
    "    tape_prop = load_tape_prop(tmp_dir, out_dir, celltype_names)\n",
    "    scaden_prop = load_scaden_prop(tmp_dir, out_dir, celltype_names)\n",
    "    message(paste0('TAPE prop Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(tape_prop)))))\n",
    "    message(paste0('Scaden prop Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(scaden_prop)))))\n",
    "\n",
    "    ### Random proportions deconvolution ### \n",
    "    message('Random deconvolution')\n",
    "    rand_prop = deconvolution_random(nnls_prop)\n",
    "    write.csv(rand_prop, paste0(out_dir, '/rand_prop.csv'), quote=F)\n",
    "    message(paste0('Random prop Accruacy as PCC: ', suppressWarnings(cor(flatten(true_prop), flatten(rand_prop)))))\n",
    "}\n",
    "\n",
    "launch_climb_only <- function(sc, bulks, true_prop, celltype_names, out_dir){ \n",
    "    ### CLIMB deconvolution ###\n",
    "    message('CLIMB deconvolution')\n",
    "    climb_prop = deconvolution_climb(sc, bulks, true_prop, celltype_names)\n",
    "    # Make sure rows/columns are in right order:\n",
    "    climb_prop = climb_prop[,colnames(true_prop)]\n",
    "    climb_prop = climb_prop[rownames(true_prop),]\n",
    "    message(paste0('CLIMB Accruacy as PCC: ', cor(flatten(true_prop), flatten(climb_prop))))\n",
    "    return(climb_prop)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "We load the data stored as `ExpressionSet` object. We will load the scRNA-seq reference dataset paired with pseudo-bulk samples made in notebook `Fig3_preproc.ipynb`. These are the list of associated single-cell reference datasets / pseudo-bulk samples with matched cell-type labels:\n",
    "\n",
    "__AML: Van Galen > Naldini cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Van Galen: `data/pseudobulks_climb/AML_vangalen_sc_es.RDS`\n",
    "- pseudo-bulks from Naldini dataset with cell-type labels transfered from Van Galen: `data/pseudobulks_climb/AML_naldini_pbulks_vg_labs.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/AML_vg2na/`\n",
    "\n",
    "__AML: Naldini > Van Galen cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Naldini: `data/pseudobulks_climb/AML_naldini_sc_es.RDS`\n",
    "- pseudo-bulks from Van Galen dataset with cell-type labels transfered from Naldini: `data/pseudobulks_climb/AML_vangalen_pbulks_na_labs.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/AML_na2vg/`\n",
    "\n",
    "__GBM: Neftel (10X) > Neftel (SmartSeq2) cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Neftel (10X): `data/pseudobulks_climb/GBM_neftel10x_sc_es.RDS`\n",
    "- pseudo-bulks from Neftel (SmartSeq2) dataset with cell-type labels transfered from Neftel (10X): `data/pseudobulks_climb/GBM_neftelSS2_pbulk_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/GBM_10x2ss2/`\n",
    "\n",
    "__GBM: Neftel (SmartSeq2) > Neftel (10X) cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Neftel(SmartSeq2): `data/pseudobulks_climb/GBM_neftelSS2_sc_es.RDS`\n",
    "- pseudo-bulks from Neftel (10X) dataset with cell-type labels transfered from Neftel(SmartSeq2): `data/pseudobulks_climb/GBM_neftel10x_pbulks_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/GBM_SS2tenX/`\n",
    "\n",
    "__Breast: Wu > Gray cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Wu: `data/pseudobulks_climb/BREAST_wu_sc_es.RDS`\n",
    "- pseudo-bulks from Gray dataset with cell-type labels transfered from Wu: `data/pseudobulks_climb/BREAST_gray_pbulks_wu_labs_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/BREAST_wu2gray/`\n",
    "\n",
    "__Breast: Gray > Wu cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Gray: `data/pseudobulks_climb/BREAST_gray_sc_es.RDS`\n",
    "- pseudo-bulks from Wu dataset with cell-type labels transfered from Gray: `data/pseudobulks_climb/BREAST_wu_pbulks_gray_labs_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/BREAST_gray2wu/`\n",
    "\n",
    "__CRC: Lee > Khaliq cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Lee: `data/pseudobulks_climb/BREAST_gray_sc_es.RDS`\n",
    "- pseudo-bulks from Khaliq dataset with cell-type labels transfered from Lee: `data/pseudobulks_climb/BREAST_wu_pbulks_gray_labs_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/CRC_lee2kha/`\n",
    "\n",
    "__CRC: Khaliq > Lee cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Khaliq: `data/pseudobulks_climb/BREAST_gray_sc_es.RDS`\n",
    "- pseudo-bulks from Lee dataset with cell-type labels transfered from Khaliq: `data/pseudobulks_climb/BREAST_wu_pbulks_gray_labs_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/CRC_kha2lee/`\n",
    "\n",
    "__MEL: Jerby > Tirosh cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Jerby: `data/pseudobulks_climb/MEL_jerby_sc_es.RDS`\n",
    "- pseudo-bulks from Tirosh dataset with cell-type labels transfered from Jerby: `data/pseudobulks_climb/MEL_tirosh_pbulks_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/MEL_jer2tir/`\n",
    "\n",
    "__MEL: Tirosh > Jerby cross-dataset pseudo-bulk deconvolution__\n",
    "\n",
    "- single-cell reference from Tirosh: `data/pseudobulks_climb/MEL_tirosh_sc_es.RDS`\n",
    "- pseudo-bulks from Jerby dataset with cell-type labels transfered from Tirosh: `data/pseudobulks_climb/MEL_jerby_pbulks_es.RDS`\n",
    "\n",
    "results stored in `data/pseudobulks_climb/MEL_tir2jer/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "singlecell_reference = \"data/pseudobulks_climb/CRC_lee_sc_es.RDS\"\n",
    "pseudobulk_target = \"data/pseudobulks_climb/CRC_khaliq_pbulks_lee_labs_es.RDS\"\n",
    "\n",
    "# Load ExpressionSet object containing scRNA-seq data from in-vitro cell mixes\n",
    "sc.es = readRDS(singlecell_reference)\n",
    "sc.es$cellType = reformat_celltypes(sc.es$cellType_original)\n",
    "celltype_names = levels(sc.es$cellType)\n",
    "# Load ExpressionSet object containing bulk RNA-seq data from in-vitro cell mixes\n",
    "bulk.es = readRDS(pseudobulk_target)\n",
    "# Load True cell line proportions (based on the volume of homogeneised cells pippeted in the mixture)\n",
    "true_prop = as.matrix(pData(bulk.es))\n",
    "colnames(true_prop) = reformat_strings(colnames(true_prop))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define output directory and temp directory\n",
    "out_dir = 'data/pseudobulks_climb/CRC_lee2kha/' ; tmp_dir = paste0(out_dir, '/tmp_tabs')\n",
    "dir.create(tmp_dir, showWarnings = FALSE, recursive = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write true proportions and celltype names in output folder for later usage\n",
    "write.csv(celltype_names, paste0(out_dir, 'celltype_names.csv'), quote=F)\n",
    "write.csv(true_prop, paste0(out_dir, 'true_prop.csv'), quote=F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cross-dataset analysis\n",
    "\n",
    "Hereafter is the script that will launch all deconvolution methods in parallel. It writes all the results in the target directory, which can then be loaded with function `load_all_props`.\n",
    "\n",
    "Warning: The next cell takes a long time to run! To run Van Galen > Naldini or Naldini > Van Galen analysis, it takes around 2/3 hours to complete.\n",
    "\n",
    "Warning2: To be able to launch TAPE/Scaden method, the function `launch_all_deconvolution_methods` calls another function, `deconvolution_scaden`, that will itself launch a bash script, `launch_TAPE_and_Scaden.sh`. This bash script has two function: 1, it will load the environment of TAPE/Scaden method, which __should already be setup before lauching `launch_all_deconvolution_methods`__. To set up this script, please refer to the README. Once the environment is set up, is can be activated by `launch_TAPE_and_Scaden.sh`, which will then launch the python script `launch_TAPE_and_Scaden.py`. Once the run is over, TAPE/Scaden table are written in the output directory and can be accessed with other proportions through `load_all_props` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Creating folders\n",
      "\n",
      "Writing tables for CiberSortX\n",
      "\n",
      "CLIMB deconvolution\n",
      "\n",
      "BayesPrism deconvolution\n",
      "\n",
      "MUSIC deconvolution\n",
      "\n",
      "NNLS deconvolution\n",
      "\n",
      "BisqueRNA deconvolution\n",
      "\n",
      "TAPE / Scaden deconvolution\n",
      "\n",
      "Launch Scaden / TAPE deconvolution\n",
      "\n",
      "TAPE prop Accruacy as PCC: 0.108634791390282\n",
      "\n",
      "Scaden prop Accruacy as PCC: 0.224150537529604\n",
      "\n",
      "Random deconvolution\n",
      "\n"
     ]
    }
   ],
   "source": [
    "launch_all_deconvolution_methods(sc = sc.es, bulks = bulk.es, true_prop = true_prop,\n",
    "                                 celltype_names = celltype_names, out_dir = out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "launch_climb_only(sc = sc.es, bulks = bulk.es, true_prop = true_prop,\n",
    "                  celltype_names = celltype_names, out_dir=out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Checking that all tables have the same cell-type labels: TRUE\n",
      "\n",
      "Checking that all tables has the same sample names: TRUE\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_props = load_all_props(out_dir, true_prop, celltype_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_res = make_all_metrics(all_props, true_prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.0985142974143845"
      ],
      "text/latex": [
       "0.0985142974143845"
      ],
      "text/markdown": [
       "0.0985142974143845"
      ],
      "text/plain": [
       "[1] 0.0985143"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "median(all_res$`CLIMB-DW`[[4]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.106910035718267"
      ],
      "text/latex": [
       "0.106910035718267"
      ],
      "text/markdown": [
       "0.106910035718267"
      ],
      "text/plain": [
       "[1] 0.10691"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "median(all_res$Scaden[[4]])"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "formats": "ipynb,md",
   "main_language": "python"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
